### PR TITLE
+str #18793 StageLogging that allows logger access in stages

### PR DESCRIPTION
**Pull request by ktoso**
#18793 was bugging me so I have it a shot.

It's more general in the sense of that other materializers can still opt-in and provide a logger if they want to. 

Also, non ActorMaterializers can opt-into providing a logger here.

### CODE REVIEW

File: akka-docs/rst/scala/stream/stream-customize.rst [hunk 1]
```diff
@@ -300,6 +300,28 @@ constructor and usually done in ``preStart``). In this case the stage **must** b
 or ``failStage(exception)``. This feature carries the risk of leaking streams and actors, therefore it should be used
 with care.
 
+Logging inside GraphStages
+--------------------------
+
+Logging debug or other important information in your stages is often a very good idea, especially when develiping
```
Comment (johanandren) [inline akka-docs/rst/scala/stream/stream-customize.rst]: develiping -> developing

File: akka-docs/rst/scala/stream/stream-customize.rst [hunk 2]
```diff
@@ -300,6 +300,28 @@ constructor and usually done in ``preStart``). In this case the stage **must** b
...
+more advances stages which may need to be debugged at some point.
+
+The helper trait ``akka.stream.stage.StageLogging`` is provided to enable you to easily obtain a ``LoggingAdapter``
+inside of a ``GraphStage`` as long as the ``Materializer`` you're using is able to provide you with a logger.
+In that sense, it serves a very similar purpose as ``ActorLogging`` does for Actors. 
+
+.. note:: 
+  Please note that you can always simply use a logging library directly inside a Stage.
+  Make sure to use an asynchronous appender however, to not accidentally block the stage when writing to files etc.
```
Comment (johanandren) [inline akka-docs/rst/scala/stream/stream-customize.rst]: Maybe also a link to the configure async logger part of logging docs?

**Review by johanandren — APPROVED** [review]
LGTM some small changes

File: akka-docs/rst/scala/stream/stream-customize.rst (hunk 2 above)
Comment (ktoso) [inline akka-docs/rst/scala/stream/stream-customize.rst]: will link to Using_the_SLF4J_API_directly

File: akka-docs/rst/scala/code/docs/stream/GraphStageLoggingDocSpec.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+  implicit val materializer = ActorMaterializer()
+  implicit val ec = system.dispatcher
+
+  //#stage-with-logging
+  final class RandomLetters extends GraphStage[SourceShape[String]] {
+    val out = Outlet[String]("MyStage.out")
+    override val shape: SourceShape[String] = SourceShape(out)
+
+    override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with StageLogging {
```
Comment (patriknw) [inline akka-docs/rst/scala/code/docs/stream/GraphStageLoggingDocSpec.scala]: override `logSource`, the anonymous inner class name is rather ugly

File: akka-stream/src/main/scala/akka/stream/stage/StageLogging.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package akka.stream.stage
+
+import akka.event.{ LoggingAdapter, NoLogging }
+import akka.stream.{ ActorMaterializer, MaterializerLoggingProvider }
+
+/**
+ * Simple way to obtain a [[LoggingAdapter]] when used together with an [[ActorMaterializer]].
+ * If used with a different materializer [[NoLogging]] will be returned.
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/stage/StageLogging.scala]: does this work with genjavadoc, or do we have to qualify the links with full package name?

**Comment by ktoso** [comment]
Addressed comments and added javadsl

**Review by patriknw — COMMENTED** [review]
looking good, need java documentation also

File: akka-docs/rst/scala/code/docs/stream/GraphStageLoggingDocSpec.scala [hunk 5]
```diff
@@ -16,22 +16,23 @@ class GraphStageLoggingDocSpec extends AkkaSpec("akka.loglevel = DEBUG") {
...
+    val out = Outlet[String]("RandomLettersSource.out")
     override val shape: SourceShape[String] = SourceShape(out)
 
-    override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with StageLogging {
-      setHandler(out, new OutHandler {
-        override def onPull(): Unit = {
-          val c = nextChar() // ASCII lower case letters
+    override def createLogic(inheritedAttributes: Attributes) =
+      new GraphStageLogic(shape) with StageLogging {
+        setHandler(out, new OutHandler {
```
Comment (patriknw) [inline akka-docs/rst/scala/code/docs/stream/GraphStageLoggingDocSpec.scala]: AbstractOutHandler

**Comment by patriknw** [comment]
perhaps you want to replace the artery StageLogging at the same time

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 6]
```diff
@@ -1231,6 +1231,12 @@ abstract class TimerGraphStageLogic(_shape: Shape) extends GraphStageLogic(_shap
 
 }
 
+/** Java API: [[GraphStageLogic]] with [[StageLogging]]. */
+abstract class GraphStageLogicWithLogging(_shape: Shape) extends GraphStageLogic(_shape) with StageLogging
+
+/** Java API: [[TimerGraphStageLogic]] with [[StageLogging]]. */
+abstract class TimerGraphStageLogicWithLogging(_shape: Shape) extends GraphStageLogic(_shape) with StageLogging
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Ouch thanks!

**Comment by ktoso** [comment]
Removed the artery logging trait

**Comment by patriknw** [comment]
seems real `GraphStageLoggingDocTest`

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
Failure was silly, no test method in junit test class (compile only test). Fixed and merging.

**Pull request merged**